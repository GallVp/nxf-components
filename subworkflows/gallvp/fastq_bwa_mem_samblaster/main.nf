include { BWA_INDEX         } from '../../../modules/gallvp/bwa/index/main'
include { BWA_MEM           } from '../../../modules/gallvp/bwa/mem/main'
include { SAMBLASTER        } from '../../../modules/gallvp/samblaster/main'

workflow FASTQ_BWA_MEM_SAMBLASTER {

    take:
    ch_fastq                // channel: [ val(meta), [ fq ] ]
    ch_reference            // channel: [ val(meta2), fasta, index ]; fasta | index

    main:
    ch_versions             = Channel.empty()

    ch_has_index            = ch_reference
                            | branch { meta2, fasta, index ->
                                yes: index
                                no: !index
                            }

    // MODULE: BWA_INDEX
    BWA_INDEX ( ch_has_index.no.map { meta2, fasta, index -> [ meta2, fasta ] } )

    ch_bwa_index            = BWA_INDEX.out.index
                            | mix(
                                ch_has_index.yes
                                | map { meta2, fasta, index ->
                                    [ meta2, index ]
                                }
                            )

    ch_versions             = ch_versions.mix(BWA_INDEX.out.versions.first())

    // MODULE: BWA_MEM
    ch_mem_inputs           = ch_fastq
                            | combine(
                                ch_bwa_index
                            )
                            | map { meta, fq, meta2, index ->
                                [ meta + [ ref_id: meta2.id ], fq, index ]
                            }

    def sort_bam            = false
    BWA_MEM(
        ch_mem_inputs.map { meta, fq, index -> [ meta, fq ] },
        ch_mem_inputs.map { meta, fq, index -> [ [], index ] },
        [ [], [] ],
        sort_bam
    )

    ch_mem_bam              = BWA_MEM.out.bam
    ch_versions             = ch_versions.mix(BWA_MEM.out.versions.first())

    // MODULE: SAMBLASTER
    SAMBLASTER ( ch_mem_bam )

    ch_blasted_bam          = SAMBLASTER.out.bam
    ch_versions             = ch_versions.mix(SAMBLASTER.out.versions.first())

    emit:
    bam                     = SAMBLASTER.out.bam    // channel: [ val(meta), bam ]
    versions                = ch_versions           // channel: [ versions.yml ]
}
