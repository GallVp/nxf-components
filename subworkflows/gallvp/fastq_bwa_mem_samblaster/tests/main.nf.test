nextflow_workflow {

    name "Test Subworkflow FASTQ_BWA_MEM_SAMBLASTER"
    script "../main.nf"
    workflow "FASTQ_BWA_MEM_SAMBLASTER"
    config './nextflow.config'

    tag "subworkflows"
    tag "subworkflows_gallvp"
    tag "subworkflows/fastq_bwa_mem_samblaster"
    tag "samblaster"
    tag "bwa/index"
    tag "bwa/mem"


    test("sarscov2-fq-gz") {

        when {
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test' ],
                        [
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true),
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkIfExists: true)
                        ]
                    ]
                )
                input[1] = Channel.of(
                    [ [ id: 'genome' ], file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true), [] ]
                )
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }

    test("sarscov2-fq-gz-stub") {

        options '-stub'

        when {
            workflow {
                """
                input[0] = Channel.of(
                    [
                        [ id:'test' ],
                        [
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_1.fastq.gz', checkIfExists: true),
                            file(params.modules_testdata_base_path + 'genomics/sarscov2/illumina/fastq/test_2.fastq.gz', checkIfExists: true)
                        ]
                    ]
                )
                input[1] = Channel.of(
                    [ [ id: 'genome' ], file(params.modules_testdata_base_path + 'genomics/sarscov2/genome/genome.fasta', checkIfExists: true), [] ]
                )
                """
            }
        }

        then {
            assertAll(
                { assert workflow.success},
                { assert snapshot(workflow.out).match()}
            )
        }
    }
}