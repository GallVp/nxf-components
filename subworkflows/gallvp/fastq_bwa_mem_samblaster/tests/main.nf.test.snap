{
    "sarscov2-fq-gz": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "ref_id": "genome"
                        },
                        "test.on.genome.samblaster.bam:md5,38ade4c8ae74d49a197286e09a6cb3ad"
                    ]
                ],
                "1": [
                    "versions.yml:md5,3b59736950270bca59c29afeb5289b4f",
                    "versions.yml:md5,80949a94ee1be8238b1c79d89ed9fa83",
                    "versions.yml:md5,ddd810edeee1008ec55b628e3a9d6204"
                ],
                "bam": [
                    [
                        {
                            "id": "test",
                            "ref_id": "genome"
                        },
                        "test.on.genome.samblaster.bam:md5,38ade4c8ae74d49a197286e09a6cb3ad"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,3b59736950270bca59c29afeb5289b4f",
                    "versions.yml:md5,80949a94ee1be8238b1c79d89ed9fa83",
                    "versions.yml:md5,ddd810edeee1008ec55b628e3a9d6204"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-22T15:55:44.323082"
    },
    "sarscov2-fq-gz-stub": {
        "content": [
            {
                "0": [
                    [
                        {
                            "id": "test",
                            "ref_id": "genome"
                        },
                        "test.on.genome.samblaster.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "1": [
                    "versions.yml:md5,3b59736950270bca59c29afeb5289b4f",
                    "versions.yml:md5,80949a94ee1be8238b1c79d89ed9fa83",
                    "versions.yml:md5,ddd810edeee1008ec55b628e3a9d6204"
                ],
                "bam": [
                    [
                        {
                            "id": "test",
                            "ref_id": "genome"
                        },
                        "test.on.genome.samblaster.bam:md5,d41d8cd98f00b204e9800998ecf8427e"
                    ]
                ],
                "versions": [
                    "versions.yml:md5,3b59736950270bca59c29afeb5289b4f",
                    "versions.yml:md5,80949a94ee1be8238b1c79d89ed9fa83",
                    "versions.yml:md5,ddd810edeee1008ec55b628e3a9d6204"
                ]
            }
        ],
        "meta": {
            "nf-test": "0.8.4",
            "nextflow": "23.10.1"
        },
        "timestamp": "2024-05-22T15:55:50.527259"
    }
}